packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/graph_based_core.rs

```diff
@@ -208,4 +526,191 @@ impl Core {
             }
         }
     }
+
+    /// Add component `child_idx` as the first child of `parent_idx`.
+    fn prepend_child_to_structure_graph(
+        &mut self,
+        parent_idx: ComponentIdx,
+        child_idx: ComponentIdx,
+    ) {
+        let parent_node = GraphNode::Component(parent_idx);
+        let child_node = GraphNode::Component(child_idx);
+
+        let children_virtual_node = self.structure_graph.get_nth_child(&parent_node, 0).unwrap();
+        self.structure_graph
+            .prepend_edge(&children_virtual_node, &child_node);
+    }
+
+    /// Add to `structure_graph` the relationships from component `component_idx` extending component `referent_idx`.
+    /// - The children of `referent` become the first children of `component` (i.e., become _virtual children_)
+    /// - Attributes of `referent` that match attributes of `component` become backup attributes of `component`,
+    ///   (i.e., become _virtual_attributes_)
+    /// - If the component types of `component` and `referent` are the same,
+    ///   then state props of `referent` are used for the state props of `component`
+    fn add_component_extending_structure(
+        &mut self,
+        component_idx: ComponentIdx,
+        referent_idx: ComponentIdx,
+    ) {
+        let component_node = GraphNode::Component(component_idx);
+        let referent_node = GraphNode::Component(referent_idx);
+
+        // The virtual node corresponding to `referent`'s children is the first child of `component`,
+        // so that `referent`'s children will be the first children of `component`.
+        let component_children_virtual_node = self
+            .structure_graph
+            .get_nth_child(&component_node, 0)
+            .unwrap();
+        let referent_children_virtual_node = self
+            .structure_graph
+            .get_nth_child(&referent_node, 0)
+            .unwrap();
```

Comment:
We've done this at least 3 times. Is it worth adding a function to `DirectedGraph` a `get_child_virtual_node` function?

Reply:
Added: `get_component_child_virtual_node()`.